src/Draco.Compiler.Api/Syntax/ParseTree.cs

```diff
@@ -42,10 +46,234 @@ public abstract partial class ParseTree
     /// </summary>
     public ParseTree? Parent { get; }
 
-    public override string ToString() => Internal.Syntax.CodeParseTreePrinter.Print(this.Green);
+    /// <summary>
+    /// The text this node was parsed from.
+    /// </summary>
+    public string TextIncludingTrivia => Internal.Syntax.CodeParseTreePrinter.Print(this.Green);
+
+    private Position? position;
+    /// <summary>
+    /// The position of the start of this node in the source text.
+    /// </summary>
+    public Position Position => this.position ??= this.ComputePosition();
+
+    private Range? range;
+    /// <summary>
+    /// The range of this node in the source text.
+    /// </summary>
+    public Range Range => this.range ??= this.ComputeRange();
+
+    /// <summary>
+    /// All <see cref="Token"/>s that this subtree consists of.
+    /// </summary>
+    public IEnumerable<Token> Tokens => this.GetTokens();
+
+    /// <summary>
+    /// Retrieves all syntactic <see cref="Diagnostic"/>s within this tree.
+    /// </summary>
+    /// <returns>The diagnostics inside this tree.</returns>
+    public IEnumerable<Diagnostic> GetAllDiagnostics() =>
+        this.CollectAllDiagnostics();
+
+    public override string ToString() => this.ComputeTextWithoutSurroundingTrivia();
     public string ToDebugString() => Internal.Syntax.DebugParseTreePrinter.Print(this.Green);
 }
 
+public abstract partial class ParseTree
```

Comment:
We have two partial of the same class in the same file ?

Reply:
Easy separability by functionality.